### DISCUSSION TITLE
PyMuPDF extracts invisible characters

### DISCUSSION THREAD

**Discussion by sergenti** [pymupdf/PyMuPDF#2263] [Q&A]
# 🐛 Describe the bug 
Hey, it seems that in one of my docs, the library is extracting invisible characters/characters of the same colors as the background.
Is there a way to parse ONLY visible chars / all characters that have a different color from the BG?

# 👨‍🔬 To Reproduce
I'd love to upload the file as well but it's not for general distribution.
If you give me a way to check if it's the former or the latter, I'll gladly check out and tell you more!

# 🖼️ Screenshots
green area (highlighted)
![Screenshot from 2023-02-27 01-25-14](https://user-images.githubusercontent.com/55017307/221446864-0ae6f345-72d3-455a-b25a-c49f4ce70e9e.png)

# ⚙️ Configuration 
Linux - Ubuntu 22.04.2 LTS (Jammy Jellyfish)
3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0] 
PyMuPDF 1.21.1: Python bindings for the MuPDF 1.21.1 library.
Version date: 2022-12-13 00:00:01.
Built for Python 3.10 on linux (64-bit).

**Comment by JorjMcKie** [comment]
This is typical "Discussions" item - no bug. So let me transfer this first.

**Reply by JorjMcKie** [reply]
The concept of "invisible text" may be quite complex.
* PDF directly supports invisibility via the `Tr` operator (text rendering mode). This is widely used by OCR engines.
* Text may simply be "behind" other objects like (intransparent) images or drawings (vector graphics).
* Text may be covered by an annotation - which in some sense is a "temporary" (= deletable) object.
* Even if text is underneath an image, that image may be transparent ... so what's the answer? Or a drawing / annotation uses a "blend mode" that keeps objects of a darker color underneath visible ***although*** it is *per se* intransparent.
* Your case: a snow rabbit in Norway.

So there is no simple way tro answer the question "Is this text / image / other object visible?"

But PyMuPDF lets you determine if a given rectangle is unicolor. Once you have extracted the bbox of text, make a Pixmap of that bbox and check if it is unicolor - or even if the dominant color is present in more than x% of all pixels.
```python
# bbox is the text boundary box
pix = page.get_pixmap(clip=bbox)
if pix.is_unicolor:
    print(f"text is invisible!")
# the following method even returns the most used color:
pix.color_topusage()
```
![grafik](https://user-images.githubusercontent.com/8290722/221519475-6e822945-0f9f-431b-a882-5d4638186300.png)
